catkin_ws/src/calqui/scripts/goto_server.py

```diff
@@ -0,0 +1,138 @@
+#!/usr/bin/env python
+
+import rospy
+import actionlib
+import numpy
+import math
+
+from calqui_msgs.msg import GoToPlaceAction
+from calqui_msgs.msg import GoToPlaceGoal
+from calqui_msgs.msg import GoToPlaceResult
+from calqui_msgs.msg import GoToPlaceFeedback
+from geometry_msgs.msg import Twist
+from geometry_msgs.msg import Pose2D
+from nav_msgs.msg import Odometry
+from tf.transformations import euler_from_quaternion, quaternion_from_euler
+
+class GoToServer:
+    def __init__(self):
+        rospy.on_shutdown(self.on_shutdown)
+        self._as = actionlib.SimpleActionServer('/goto_place', 
+                GoToPlaceAction, execute_cb=self.on_goal, auto_start=False)
+        self._vel_pub = rospy.Publisher("/cmd_vel", Twist, queue_size=10)
+        self._fb_pub = rospy.Publisher("/goto_place/feedback", GoToPlaceFeedback, queue_size=10)
+        self._odom_sub = rospy.Subscriber("/odom", Odometry, self.on_odom)
+        self._current_pose = Pose2D()
+        self._goal_pose = Pose2D()
+        self._as.start()
+        
+        rospy.loginfo('Simple Action Server has been started')
+
+    def on_shutdown(self):
+        rospy.loginfo('stopping vel')  
+        self.set_vel(0,0)
+
+    def on_odom(self, msg):
+        global roll, pitch, yaw
+        orientation_q = msg.pose.pose.orientation
+        orientation_list = [orientation_q.x, orientation_q.y, orientation_q.z, orientation_q.w]
+        (roll, pitch, yaw) = euler_from_quaternion (orientation_list)
+        # rospy.loginfo('yaw:{}, x:{} y:{} z:{}'.format(round(yaw,2), round(msg.pose.pose.position.x,2),round(msg.pose.pose.position.y,2),round(msg.pose.pose.position.z,2) ))
+        self._current_pose.x = msg.pose.pose.position.x
+        self._current_pose.y = msg.pose.pose.position.y
+        self._current_pose.theta = yaw
+        
+        # rospy.loginfo('current pose is {}'.format(self._current_pose))
+        fb = GoToPlaceFeedback()
+        fb.distance_to_goal = self.distance_to_goal()
+        self._fb_pub.publish(fb)
+        
+
+    def set_vel(self, linear_x, angular_z):
+        rospy.loginfo('publishing vel')
+        msg = Twist()
+        msg.linear.x = linear_x
+        msg.linear.y = 0
+        msg.linear.z = 0
+        msg.angular.x = 0
+        msg.angular.y = 0
+        msg.angular.z = angular_z
+        self._vel_pub.publish(msg)
+
+    def distance_to_goal(self):
+        return math.sqrt(math.pow((self._goal_pose.y - self._current_pose.y),2) + math.pow((self._goal_pose.x - self._current_pose.x),2))
+
+    # absolute angle to the goal
+    def angle_to_goal(self):
+        # avoid division by 0
+        if self._current_pose.x == 0:
+            return 0 
+        else:
+            return numpy.arctan2((self._goal_pose.y - self._current_pose.y), (self._goal_pose.x - self._current_pose.x))
+
+    # how far are we of pointing to the goal?
+    def deviation(self):
+        return self._current_pose.theta - self.angle_to_goal()
```

Comment:
you should wrap the angle between -pi/pi, to avoid the robot adjusting the deviation in the incorrect direction.
See https://numpy.org/doc/stable/reference/generated/numpy.unwrap.html.